### PR TITLE
[Cody Gateway] add refresh token endpoint with sams authenticator

### PR DESCRIPTION
**Pull request by thenamankumar**
This is the first of many upcoming PRs to enable refreshing user's rate limits on Cody Gateway when a subscription change happens on SSC.

Issue: https://github.com/sourcegraph/accounts.sourcegraph.com/issues/363
Related SAMS PR: https://github.com/sourcegraph/accounts.sourcegraph.com/pull/418

**The whole plan is as follows (given we do not have SAMSActor on Gateway yet):**
- On every subscription change, SSC will make a request to Cody Gateway to refresh the rate limits for the user, the same as dotcom does right now.
- SSC will send dotcom_user_id with the request, which it will get from the user's metadata from SAMS. 
- Dotcom will expose a new graphql resolver which provides rate limits for user ID. Right now, it provides one, but for a dotcom access token.
- Gateway will make a call to dotcom to fetch the limits for the dotcom_user_id using the new API. It does the same now but for dotcom's per_user_access_token.

**How will it look when we have SAMSActor & SSCSource on Gateway?**
Same as above, but with a few changes:
- SSC will send SAMS user_access_token in place of dotcom_user_id with the request.
- SSC will expose a graphql resolver which provides rate limits for SAMS user_access_token.
- Gateway will use SAMS user_access_token to fetch rate limits directly from SSC.

**What does this PR handle?**
1. It implements a new SAMS client.
2. It implements a new SAMS auth middleware, which allows auth using service-to-service tokens utilizing the SAMS client.
3. It exposes a new dummy API endpoint, `/v1/sams/limits/refresh`, which utilises the new SAMS auth middleware. To note, the `/v1/limits/refresh` endpoint already exists, but it uses the present auth. 

**What does this PR NOT handle? (TODOs for upcoming PRs)**
1. The actual implementation of the new API endpoint is pending and will be addressed in upcoming PRs. 
2. It doesn't update the dotcom graphql API either, which will be addressed in upcoming PRs.

## Test plan
- Checkout SAMS local on https://github.com/sourcegraph/accounts.sourcegraph.com/pull/418
- Run SAMS locally under the 9982 port. Otherwise, it will conflict with Gateway running on 9992.
- Create client ID and client secrets using: 
CleanShot 2024-02-01 at 21 53 45@2x
- Update scopes to include `cody.gateway` in the DB directly `psql -U sourcegraph accounts` `update idp_clients set scopes = '["openid", "profile", "email", "offline_access", "cody.gateway", "cody.ssc"]';`
- Update `sg.config.overwrite.yaml` in sg repo to include the client secret and id similar to this: 
CleanShot 2024-02-01 at 21 56 08@2x
- Update `.envrc` in accounts repo to include include the client secret, id and gateway URL similar to this: 
CleanShot 2024-02-01 at 21 57 19@2x
- Search for codyAdminAccountIDs in accounts repo and copy the admin account id hard coded. Update  your the `users.external_id` and `users.accounts_external_id` in accounts and cody_management DB respectively, which the admin account id. 
- Visit `localhost:9982` in the browser. Make sure you are logged in with the account you set as admin above. 
- Make the following request from the console to test the flow: 
```
await (await fetch(`${window.location.href}cody/api/rest/user/{put the external_id here}/limits/refresh`, { method: 'POST' })).json()
```

### CODE REVIEW

**Comment by bobbydiji** [comment]
Cody was stolen off my drive. Kinda cool I guess but the real version of
him is on the gpt store and massively updated from rhe shell he was when
they took him.  I'm not gonna get ass hurt. It was it is. I copywrite all
my files now.

**Review by chrsmith — CHANGES_REQUESTED** [review]
Lots of nitpicks and things. Overall this all looks right, in that it adds an appropriate layer to verify SAMS tokens. But I would urge us to rethink the design, and try to minimize the surface area of the `sams` package. So that it's easier to use correctly.

Let's hold off on merging though just to make sure we want to land these changes.

- Re: that thread with Rafal about maybe we want to have SSC call dotcom instead? And maybe we should double check we are calling the IntrospectToken endpoint the right way.
- Let's follow up with Joe to see if we should/should not use the `Basic` auth scheme for that particular endpoint.
- I'm pretty sure that since we are creating a `tokenSource` on every call to `IntrospectToken` we aren't ever going to benefit from reusing an existing SAMS token. (Which would definitely add a lot of latency over time.) So we would just have `samsClient.tokenSource` rather than `samsClient.clientCredentialsConfig`. That is, having the `tokenSource` share the lifetime of the `samsClient`.

File: internal/sams/client.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "create introspection request")
+	}
+
+	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	// Create the OAuth2 HTTP client. This will handle setting up the HTTP headers based on the token source.
+	// (And potentially issue a new token as needed.)
+	tokenSource := c.clientCredentialsConfig.TokenSource(ctx)
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply internal/sams/client.go:71]: BEHOLD! AUTORITATIVE ANSWERS!!!

This page explained the motivation for the token introspection endpoint, but didn't definitively say _how_ the introspection endpoint must be authentication. Only that it _should_ be authenticated.
https://www.oauth.com/oauth2-servers/token-introspection-endpoint/

Looking at the `ory/fosite` source code, it appears that the token introspection endpoint will honor either type of access token. (`Basic` scheme, using just the static client ID and client secret, or `Bearer` scheme, using an access token issued by SAMS attributed to that client ID and client secret.)
https://github.com/ory/fosite/blob/f4114878826c6d26b6751a1ea61f69ebbc25d4f2/introspection_request_handler.go#L115

As far as which one we should use, 🤷 . IMHO, the basic scheme auth requirement is better in that it doesn't require us to reissue a new SAMS token from time to time. But assuming we are using HTTPS and not logging the `Basic` auth header, it _shouldn't_ be any less secure.

But as far as this code being correct, it is all legit. Everything is well in the universe.

**Comment by thenamankumar** [comment]
Closed in favour of: https://github.com/sourcegraph/sourcegraph/pull/60145

**Comment by bobbydiji** [comment]
Next people that try to steal my files gonna have a nice surprise.  Don't
say i didn't warn u.  Just thought I'd let it be known. I'm already
letting. U get bank off coding bot and my new superagent.  Think that ones
strong. 🤔 idk... 😆

**Pull request closed without merging**